Adding test to disprove #5584

lorenzo opened the pull request:
Added the test proposed in #5584 but everything passes as expected.

lorenzo committed:
Adding test to disprove #5584

lorenzo committed:
Corrected casing

ndm2 commented:
Interesting... have there been any recent changes that would affect this? I could swear that the test failed before, but now it's working fine :confused:

lorenzo committed:
Fixed CS error

lorenzo commented:
Nothing has changed in the ORM since RC1 was released @ndm2

ndm2 commented:
That's weird... could you give the following a try

```php
$comments = TableRegistry::get('Comments');
$comments->belongsTo('Articles');

$articles = TableRegistry::get('Articles');
$articles->belongsTo('Authors');
$articles->belongsToMany('Tags');

$result = $comments
	->find()
	->matching('Articles.Tags', function($q) {
		return $q->where(['Tags.id' => 2]);
	})
	->contain('Articles.Authors')
	->first();

$this->assertNotNull($result->article->author);
```

For me this is currently failing like the original test did before, and it is fixable as described in #5584, ie by adding a matching for the association (`matching('Articles.Authors')`) or by removing the existing matching.

lorenzo committed:
Starting to use an object for representing an eagerloadable

lorenzo committed:
Fixing most of the failing test

lorenzo committed:
Fixed EagerLoader tests

lorenzo committed:
Fixed another test by fixing a mistake in previous commits

lorenzo committed:
Making sure contain and matching clashes are always resolved, fixes #5584

lorenzo committed:
Refactoring code so strategies are only corrected at the end

lorenzo committed:
Cleaning up the code and fixing some doc blocks

lorenzo committed:
Adding doc blocks to EagerLoadable

lorenzo committed:
Merge branch '3.0-matching-fixes' of github.com:cakephp/cakephp into 3.0-matching-fixes

lorenzo commented:
@ndm2 Thanks for the new test case, it helped me find the issue.

@markstory I had to do yet another surgery to `EagerLoader`. In this case I finally went with using objects instead of array references as it was getting ridiculous and because after all that work that made the fix for this trivial.

lorenzo commented on src/ORM/EagerLoader.php:
```diff
@@ -445,19 +452,24 @@ protected function _correctStrategy(&$config, $alias)
             }
+
+            if ($inMatching) {
+                $this->_correctStrategy($loadable, $table);
```
This line was the actual fix

lorenzo committed:
Fixing CS errors

markstory committed:
Doc block updates.

Make Eagerloadable internal. End developers do not directly interact
with this class, as it is mostly an implementation detail of the
eagerloader.

markstory commented on src/ORM/EagerLoader.php:
```diff
@@ -369,18 +374,22 @@ protected function &_normalizeContain(Table $parent, $alias, $options, $paths)
+        $eagerLoadable = new EagerLoadable($alias, $config);
 
         if ($config['canBeJoined']) {
-            $this->_aliasList[$paths['root']][$alias][] =& $config;
```
I like that the reference operators are going away :clap:

markstory commented on src/ORM/EagerLoader.php:
```diff
@@ -412,26 +421,24 @@ protected function _fixStrategies()
-     * but is not possible to change the strategy due to conflicting settings
+     * @return void
      */
-    protected function _correctStrategy(&$config, $alias)
```
Looks like the $alias parameter is unused from this diff.

markstory committed:
Remove unused argument.

markstory commented on src/ORM/EagerLoadable.php:
```diff
@@ -0,0 +1,268 @@
+        if ($path === null) {
+            return $this->_propertyPath;
+        }
+        $this->_propertyPath = $path;
```
Is there any reason for these properties to be mutable? Outside of canBeJoined & config none of the properties are ever mutated. It might be simpler to have a value object.

lorenzo replied:
I started the class thinking I would use the setters, but I guess I can remove them now :)

lorenzo committed:
Removing setters for prperties in EagerLoadable, fixing indentation

Also removed a parameter in a function that is not used anymore

lorenzo committed:
Merge branch '3.0-matching-fixes' of github.com:cakephp/cakephp into 3.0-matching-fixes

Conflicts:
	src/ORM/EagerLoadable.php
	src/ORM/EagerLoader.php

lorenzo commented:
I took care of the suggestion and will merge this now. Please let me know if there was anything else you wanted changed.
